### PR TITLE
sl-90: search functionality for sites

### PR DESCRIPTION
**Pull request by vbazhynov**
all cards
image
after search
image
postman
image

### CODE REVIEW

File: frontend/src/pages/sites/sites.tsx [hunk 1]
```diff
@@ -1,16 +1,29 @@
-import { Button, Loader, PageLayout } from '~/libs/components/components.js';
+import React, { useCallback } from 'react';
+import { Link } from 'react-router-dom';
```
Comment (jaroslawtt) [inline frontend/src/pages/sites/sites.tsx]:
```suggestion

```

**Review by jaroslawtt — CHANGES_REQUESTED** [review]

File: backend/src/packages/sites/site.repository.ts [hunk 2]
```diff
@@ -34,6 +46,15 @@ class SiteRepository
 
     return SiteEntity.initialize(site);
   }
+
+  // private getWhereName =
+  //   (query: string | undefined) =>
+  //   (builder: QueryBuilder<>): void => {
```
Comment (andanf-e) [inline backend/src/packages/sites/site.repository.ts]: FYI about commented code

File: frontend/src/libs/components/input/styles.module.scss [hunk 3]
```diff
@@ -18,6 +18,20 @@
...
+  width: 100%;
+  height: 36px;
+  padding: 6px 8px 6px 32px;
+  color: var(--gray-600);
+  font-size: 14px;
+  line-height: 24px;
+  background-color: var(--white);
+  border: 2px solid var(--black);
+  outline: none;
+}
```
Comment (andanf-e) [inline frontend/src/libs/components/input/styles.module.scss]: The only this that was changed from `.input` class is a padding - why do we need to duplicate the styles?

File: frontend/src/libs/components/input/styles.module.scss [hunk 4]
```diff
@@ -51,3 +65,10 @@
   display: block;
   height: 20px;
 }
+
+.search-input-icon {
+  position: absolute;
```
Comment (andanf-e) [inline frontend/src/libs/components/input/styles.module.scss]: Can't find a relative el to this absolute icon

File: frontend/src/libs/components/input/input.tsx [hunk 5]
```diff
@@ -48,10 +52,12 @@ const Input = <T extends FieldValues>({
       >
         {label}
       </span>
+      {icon && <Icon iconName={icon} className={styles['search-input-icon']} />}
```
Comment (andanf-e) [inline frontend/src/libs/components/input/input.tsx]: What if passed icon is not a `search` Icon? why do we need to specify the className with this relation?

File: frontend/src/packages/sites/sites-api.ts [hunk 6]
```diff
@@ -44,14 +45,16 @@ class SitesApi extends HttpApi {
...
+    parameters,
+  }: SiteGetByProjectParametersDto): Promise<SiteGetAllResponseDto> {
     const response = await this.load(
       this.getFullEndpoint(
-        configureString(SitesApiPath.PROJECT_$PROJECT_ID, {
-          projectId,
+        configureString(SitesApiPath.PROJECT_$PROJECT_ID, SitesApiPath.ROOT, {
+          projectId: projectId.toString(),
         }),
+        parameters.pattern ? `?pattern=${parameters.pattern}` : '',
```
Comment (andanf-e) [inline frontend/src/packages/sites/sites-api.ts]: What would be with chars that should be escaped?

File: frontend/src/pages/sites/sites.tsx (hunk 1 above)
Comment (vbazhynov) [reply frontend/src/pages/sites/sites.tsx]: +

**Review by what1s1ove — CHANGES_REQUESTED** [review]
update the branch with `development`

try to implement everything as in the project search functionality

File: backend/src/packages/sites/site.controller.ts [hunk 7]
```diff
@@ -67,15 +69,17 @@ class SiteController extends Controller {
     this.siteService = siteService;
 
     this.addRoute({
-      path: SitesApiPath.PROJECT_$PROJECT_ID,
+      path: `${SitesApiPath.PROJECT_$PROJECT_ID}${SitesApiPath.ROOT}`,
```
Comment (what1s1ove) [inline backend/src/packages/sites/site.controller.ts]: not sure that you need it

File: shared/src/packages/sites/libs/types/sites-search-request-dto.type.ts [hunk 8]
```diff
@@ -0,0 +1,5 @@
+type SitesFilterQueryDto = {
+  pattern: string;
```
Comment (what1s1ove) [inline shared/src/packages/sites/libs/types/sites-search-request-dto.type.ts]:
```suggestion
  name: string;
```

File: shared/src/packages/sites/libs/validation-schemas/site-get-by-project.validation-schema.ts [hunk 9]
```diff
@@ -1,12 +1,14 @@
...
 
 import { type SiteGetByProjectParametersDto } from '../types/types.js';
+import { siteSearch } from './site-search.validation-schema.js';
 
 const siteGetByProjectParameters = joi.object<
   SiteGetByProjectParametersDto,
   true
 >({
   projectId: joi.number().required(),
+  parameters: siteSearch,
```
Comment (what1s1ove) [inline shared/src/packages/sites/libs/validation-schemas/site-get-by-project.validation-schema.ts]:
```suggestion
```
it should be a query parameters. not just a parameters

File: shared/src/packages/sites/libs/validation-schemas/site-search.validation-schema.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import joi from 'joi';
+
+import { type SitesFilterQueryDto } from '../types/types.js';
+
+const siteSearch = joi.object<SitesFilterQueryDto, true>({
```
Comment (what1s1ove) [inline shared/src/packages/sites/libs/validation-schemas/site-search.validation-schema.ts]: do the same name as for project

File: frontend/src/pages/sites/sites.tsx [hunk 11]
```diff
@@ -30,6 +46,27 @@ const Sites: React.FC = () => {
     status: sites.dataStatus,
   }));
 
+  const { control, errors, handleSubmit } = useAppForm<SitesFilterQueryDto>({
+    defaultValues: { pattern: '' },
+    validationSchema: sitesFilterValidationSchema,
+  });
+
+  const onInputChange = useCallback(
```
Comment (what1s1ove) [inline frontend/src/pages/sites/sites.tsx]: **A9**

File: frontend/src/libs/helpers/debounce/debounce.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
+type FunctionToDebounce<T extends unknown[], R> = (...arguments_: T) => R;
+
+function debounce<T extends unknown[], R>(
```
Comment (what1s1ove) [inline frontend/src/libs/helpers/debounce/debounce.ts]: you have it in the shared folder

File: frontend/src/libs/components/input/input.tsx [hunk 13]
```diff
@@ -15,7 +15,7 @@ import styles from './styles.module.scss';
 type Properties<T extends FieldValues> = {
   control: Control<T, null>;
   errors: FieldErrors<T>;
-  label: string;
+  label?: string;
```
Comment (what1s1ove) [inline frontend/src/libs/components/input/input.tsx]:
```suggestion
  label: string;
```
use `isLabelVisuallyHidden` if you need to hide the label

**Review by what1s1ove — CHANGES_REQUESTED** [review]
implement most of functional the same as in project (type/parameters, etc names)

File: frontend/src/assets/img/input-search.svg
Comment (what1s1ove) [inline frontend/src/assets/img/input-search.svg:1]: remove this icon

File: frontend/src/packages/sites/sites-api.ts [hunk 14]
```diff
@@ -44,14 +46,28 @@ class SitesApi extends HttpApi {
...
+    {
+      return Object.values(queryParameters).some((element) => element !== '');
+    }
+  }
+
+  public async getByProjectId({
+    projectId,
+    queryParameters,
+  }: SiteGetByProjectParametersDto): Promise<SiteGetAllResponseDto> {
+    const searchParameters = this.isQueryParametersExist(queryParameters)
```
Comment (what1s1ove) [inline frontend/src/packages/sites/sites-api.ts]: do not need this an additional message

File: frontend/src/pages/sites/sites.tsx (hunk 1 above)
Comment (what1s1ove) [reply frontend/src/pages/sites/sites.tsx]: unresolved

File: shared/src/packages/sites/libs/validation-schemas/site-get-by-project.validation-schema.ts [hunk 15]
```diff
@@ -1,12 +1,14 @@
...
 
 import { type SiteGetByProjectParametersDto } from '../types/types.js';
+import { siteFilter } from './site-filter.validation-schema.js';
 
 const siteGetByProjectParameters = joi.object<
   SiteGetByProjectParametersDto,
   true
 >({
   projectId: joi.number().required(),
+  queryParameters: siteFilter,
```
Comment (what1s1ove) [inline shared/src/packages/sites/libs/validation-schemas/site-get-by-project.validation-schema.ts]:
```suggestion\
```
it should be here. it validation for parameters

File: shared/src/packages/sites/libs/types/sites-get-all-parameters-dto.type.ts [hunk 16]
```diff
@@ -0,0 +1,5 @@
+type SitesGetAllParametersDto = {
```
Comment (what1s1ove) [inline shared/src/packages/sites/libs/types/sites-get-all-parameters-dto.type.ts]: not sure that you use this type

File: frontend/src/packages/sites/sites-api.ts (hunk 14 above)
Comment (andanf-e) [inline frontend/src/packages/sites/sites-api.ts]: Not sure if you need a helper for forming search params, check out how it was done on `getProjects`

File: frontend/src/pages/sites/sites.tsx [hunk 17]
```diff
@@ -1,16 +1,29 @@
...
-import { configureString } from '~/libs/helpers/helpers.js';
+import { configureString, initDebounce } from '~/libs/helpers/helpers.js';
 import {
-  useAppDispatch,
+  useAppForm,
   useAppSelector,
   useEffect,
   useParams,
 } from '~/libs/hooks/hooks.js';
+import { useAppDispatch } from '~/libs/hooks/use-app-dispatch/use-app-dispatch.hook.js';
```
Comment (what1s1ove) [inline frontend/src/pages/sites/sites.tsx]: use proxy for imports. fix for all
```suggestion
import { useAppDispatch } from '~/libs/hooks/hooks.js';
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: shared/src/packages/sites/libs/types/sites-get-all-parameters-dto.type.ts (hunk 16 above)
Comment (vbazhynov) [reply shared/src/packages/sites/libs/types/sites-get-all-parameters-dto.type.ts]: i did it like in Search Projects,  for me it strange why do we need two different types, one on the page and other in slices and backend. but there it done like this
image

File: shared/src/packages/sites/libs/validation-schemas/site-get-by-project.validation-schema.ts (hunk 15 above)
Comment (vbazhynov) [reply shared/src/packages/sites/libs/validation-schemas/site-get-by-project.validation-schema.ts]: you mean to do all validation in one file?  delete site-filter.validation-schema.ts and do yt all here?

Comment (what1s1ove) [reply shared/src/packages/sites/libs/validation-schemas/site-get-by-project.validation-schema.ts]: you should have 2 separate validation
1 for parameters
2 for query parameters

File: shared/src/packages/sites/libs/types/sites-get-all-parameters-dto.type.ts (hunk 16 above)
Comment (what1s1ove) [reply shared/src/packages/sites/libs/types/sites-get-all-parameters-dto.type.ts]: `/projects/:id`  – parameters
`/projects?name=supper` – query parameters

Comment (vbazhynov) [reply shared/src/packages/sites/libs/types/sites-get-all-parameters-dto.type.ts]: http://localhost:3000/api/v1/sites/project/2?name=vla here is our search string, 
/projects?name=supper –  we need to find not from projects sites but all sites?

Comment (vbazhynov) [reply shared/src/packages/sites/libs/types/sites-get-all-parameters-dto.type.ts]: or you mean to crate first class for ID, scond for searc parameters and third to connect it in action?

File: frontend/src/packages/sites/sites-api.ts (hunk 14 above)
Comment (vbazhynov) [reply frontend/src/packages/sites/sites-api.ts]: http://localhost:3000/api/v1/sites/project/2?name= 
even when we don`t have any search parameters query is present in this case, it`s ok?

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**